### PR TITLE
adding azure openai compatibility.

### PR DESCRIPTION
**Pull request by copypastedeveloper**
I have been using this library to work with openai but need to transition to Azure openai instead.  The only real changes are the endpoints are formatted a little differently and auth is a little different as well.  everything here should be completely backwards compatible.  As far as I know once this is merged this will be the first C# library to support azure openai.

### CODE REVIEW

**Comment by kayhantolga** [comment]
Thanks for the PR, I will have a look. I have been trying to get access to Azure OpenAI but I couldn't succeed yet. Did you find a chance to test your changes?

**Comment by kayhantolga** [comment]
also, could you change your target branch to dev, please?

**Comment by copypastedeveloper** [comment]
> Thanks for the PR, I will have a look. I have been trying to get access to Azure OpenAI but I couldn't succeed yet. Did you find a chance to test your changes?

I did, it all works and I was impressed with the ease of making the changes.  If you'd like an in I can make an intro at microsoft to get you pushed up in the queue.

**Comment by kayhantolga** [comment]
That would be great!

File: OpenAI.SDK/OpenAiOptions.cs [hunk 1]
```diff
@@ -21,16 +29,34 @@ public class OpenAiOptions
...
-
+        /// <summary>
+        ///     Create an instance of this class with the necessary information to connect to the azure open ai api
+        /// </summary>
+        /// <param name="baseDomain">Domain for your open ai instance</param>
+        /// <param name="deploymentId">The id of your deployment of open ai</param>
+        /// <param name="apiVersion">The azure open ai api version</param>
+        /// <param name="apiKey">Token used for authentication</param>
+        /// <returns>A valid OpenAiSettings instance configured with the method inputs</returns>
+        static OpenAiOptions CreateAzureSettings(string baseDomain, string deploymentId, string apiVersion, string apiKey)
```
Comment (kayhantolga) [inline OpenAI.SDK/OpenAiOptions.cs:48]: Why we need this?

File: OpenAI.SDK/OpenAiOptions.cs [hunk 2]
```diff
@@ -1,9 +1,17 @@
...
+    {
+        OpenAi = 1,
+        Azure = 2
+    }
+
     public class OpenAiOptions
     {
         public static readonly string SettingKey = "OpenAIServiceOptions";
 
+        public virtual ApiType ApiType { get; set; } = ApiType.OpenAi;
```
Comment (kayhantolga) [inline OpenAI.SDK/OpenAiOptions.cs:13]: Are there any specific reason to make this property Virtual?

Comment (copypastedeveloper) [reply OpenAI.SDK/OpenAiOptions.cs:13]: Nope, was originally going to make a base class and make the properties readonly, but decided not to and forgot to clean this up.  I'll fix it.

Comment (kayhantolga) [reply OpenAI.SDK/OpenAiOptions.cs:13]: please don't change it. I am going to send a commit to fix a couple of things. I am going to include this change too.

File: OpenAI.SDK/OpenAiOptions.cs (hunk 1 above)
Comment (copypastedeveloper) [reply OpenAI.SDK/OpenAiOptions.cs:48]: It isn't needed so much as it is convenient for people who don't use the appsettings based service collection extensions.  It's hard to know which things are actually required for azure vs not azure with everything being in the same settings class.  I can remove this and change the validation instead though, probably fits with the project better.

**Comment by kayhantolga** [comment]
As I understand we need to change base domain too and PR doesn't include this change, right?
`https://YOUR_RESOURCE_NAME.openai.azure.com/openai/deployments/YOUR_DEPLOYMENT_NAME/embeddings?api-version=2022-12-01`

**Comment by copypastedeveloper** [comment]
> As I understand we need to change base domain too and PR doesn't include this change, right?
> 
> `https://YOUR_RESOURCE_NAME.openai.azure.com/openai/deployments/YOUR_DEPLOYMENT_NAME/embeddings?api-version=2022-12-01`

There was already a basedomain in the settings so no change needed.

**Comment by kayhantolga** [comment]
I have just pushed some changes, and I would appreciate it if you could review it and let me know your thoughts on the changes. Please note that my Resharper reverted some of your Code Style changes. Most of the changes involve naming and code style adjustments. Additionally, I made an update to OpenAI.SDK/OpenAiOptions.cs to automatically recognize which provider the developer intends to use. 
I expect the ResourceName to be provided as an argument rather than asking for the base URL. This change should make it easier to use.

File: OpenAI.SDK/OpenAiOptions.cs [hunk 3]
```diff
@@ -22,15 +63,80 @@ public class OpenAiOptions
...
+        /// <param name="resourceName">Resource Name of your Azure OpenAI resource</param>
+        /// <param name="deploymentId">The id of your deployment of OpenAI</param>
+        /// <param name="apiVersion">The azure open ai api version</param>
+        /// <param name="apiKey">Token used for authentication</param>
+        /// <returns>A valid OpenAiSettings instance configured with the method inputs</returns>
+        private static OpenAiOptions CreateAzureSettings(string apiKey, string resourceName, string deploymentId, string? apiVersion)
+        {
+            return new OpenAiOptions
+            {
+                _providerType = ProviderType.Azure,
```
Comment (copypastedeveloper) [inline OpenAI.SDK/OpenAiOptions.cs:128]: I would think this should set the property not the field.

File: OpenAI.SDK/OpenAiOptions.cs [hunk 4]
```diff
@@ -22,15 +63,80 @@ public class OpenAiOptions
         /// </summary>
         public string ApiKey { get; set; } = null!;
 
-        public string ApiVersion { get; set; } = "v1";
+        /// <summary>
+        ///     Default Api Version
+        /// </summary>
+        public string ApiVersion
+        {
+            get
```
Comment (copypastedeveloper) [inline OpenAI.SDK/OpenAiOptions.cs:71]: I've never been a fan of property getters with logic, but it's your library.  😅

File: OpenAI.SDK/OpenAiOptions.cs [hunk 5]
```diff
@@ -22,15 +63,80 @@ public class OpenAiOptions
...
+        ///     Base Domain
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException"></exception>
+        public string BaseDomain
+        {
+            get
+            {
+                return _baseDomain ??= ProviderType switch
+                {
+                    ProviderType.OpenAi => OpenAiDefaultBaseDomain,
```
Comment (copypastedeveloper) [inline OpenAI.SDK/OpenAiOptions.cs:93]: It seems to me like having the settings behave the same for each type is more important than what you're solving for.  For reference, the way I had it originally is the way the official python library does it.

File: OpenAI.SDK/OpenAiOptions.cs [hunk 6]
```diff
@@ -1,9 +1,50 @@
...
+        private ProviderType? _providerType;
+
+        /// <summary>
+        ///     Get Provider Type
+        /// </summary>
+        public ProviderType ProviderType
+        {
+            get
+            {
+                _providerType ??= !string.IsNullOrEmpty(DeploymentId) || !string.IsNullOrEmpty(ResourceName) ? ProviderType.Azure : ProviderType.OpenAi;
```
Comment (copypastedeveloper) [inline OpenAI.SDK/OpenAiOptions.cs:42]: Same comment as below, I'd rather just make people be explicit that try and magic my way into things.

File: OpenAI.SDK/Managers/OpenAIService.cs [hunk 7]
```diff
@@ -23,15 +23,29 @@ public OpenAIService(OpenAiOptions settings, HttpClient? httpClient = null)
 
             _httpClient = httpClient ?? HttpClientFactory.Create();
             _httpClient.BaseAddress = new Uri(settings.BaseDomain);
-            var authKey = settings.ApiKey;
-            _httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {authKey}");
-            var organization = settings.Organization;
-            if (!string.IsNullOrEmpty(organization))
+            
+            switch (settings.ProviderType)
```
Comment (copypastedeveloper) [inline OpenAI.SDK/Managers/OpenAIService.cs:27]: I've never liked switch for control flow.

**Review by copypastedeveloper — COMMENTED** [review]
A few notes from my side.  I'm fine to accept most of these changes, but I'd really prefer to change the base domain back.

**Comment by kayhantolga** [comment]
I just updated the codes.
I have tried to reduce the number of logics in fields. Developers now explicitly set Provider Type to Azure if they want to use Azure Services. Otherwise, it was getting complicated to support different ways to initialize the library. But for the base domain and default Api version, I think we need to keep logic calculations, I couldn't find another simple way to bind these default values.

I have updated the validation method and CreateAzureSettings to support custom base domains. You were right if you have a custom domain you probably want to use base domain init instead of providing a resource name.

What do you think about the newer version? I am very sleepy at the moment, I might cause some small bugs. I will double-check again. Hopefully, we can release this version next week.

**Comment by copypastedeveloper** [comment]
This looks good to me and works.  I made a couple small tweaks, api version cannot ever be null or empty given the getter will throw exception, so I just called the getter in tha validator.  The changes are pretty self explanatory i think, substituting some of strings for variables, etc.  I'm looking forward to getting this mainlined and released so I can start using it in our projects.

**Pull request merged**